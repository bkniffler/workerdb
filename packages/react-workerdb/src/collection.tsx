import * as React from 'react';
import { WorkerDBCollection } from 'workerdb';
import Context from './context';
import render, { RenderProps } from './factory/render';

interface RenderPropsCollection extends RenderProps<WorkerDBCollection> {
  name: string;
}
const WorkerDBCollection = (props: RenderPropsCollection) => (
  <Context.Consumer>
    {value => render(props, { value: value && value.c(props.name) })}
  </Context.Consumer>
);

export default WorkerDBCollection;
