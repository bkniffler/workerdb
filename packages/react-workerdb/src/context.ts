import { createContext } from 'react';
import { WorkerDBClientRX } from 'workerdb';

export default createContext<WorkerDBClientRX | undefined>(undefined);
