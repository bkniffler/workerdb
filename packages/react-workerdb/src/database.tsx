import * as React from 'react';
import Context from './context';
import render, { RenderProps } from './factory/render';

const WorkerDBConsumer = (props: RenderProps) => (
  <Context.Consumer>{value => render(props, { value })}</Context.Consumer>
);

export default WorkerDBConsumer;
