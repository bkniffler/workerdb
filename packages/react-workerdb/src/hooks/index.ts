export { default as useDB } from './use-db';
export { default as useFind } from './use-find';
export { default as useFindOne } from './use-find-one';
export { default as useInsert } from './use-insert';
