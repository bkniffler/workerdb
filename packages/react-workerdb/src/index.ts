// React Core
export { default as WorkerDB } from './workerdb';
export { default as Database } from './database';
export { default as Collection } from './collection';
export { default as useCollection } from './use-collection';
export * from './hooks';

// React Sync
export { default as Syncing } from './sync/consumer';

// React Methods
export { default as FindOne } from './query/find-one';
export { default as Find } from './query/find';
export { default as Call } from './query/call';
