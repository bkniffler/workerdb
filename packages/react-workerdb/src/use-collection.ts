import * as React from 'react';
import { WorkerDBCollection, WorkerDB } from 'workerdb';
import Context from './context';

function useCollection<T = any>(name: string): WorkerDBCollection<T> {
  const value = React['useContext'](Context) as WorkerDB;
  return value && value.c(name);
}

export default useCollection;
