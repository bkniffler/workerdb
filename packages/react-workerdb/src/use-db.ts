import * as React from 'react';
import { WorkerDBClientRX } from 'workerdb';
import Context from './context';

const defaultValue = {};
function useDB(): WorkerDBClientRX {
  const value = React.useContext(Context) as WorkerDBClientRX;
  return value || defaultValue;
}

export default useDB;
