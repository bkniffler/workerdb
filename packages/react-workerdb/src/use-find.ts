import * as React from 'react';
import { WorkerDBClientRX } from 'workerdb';
import { stringify } from './utils';
import Context from './context';

interface IUseFind<T> {
  value: Array<T>;
  error: Error | null;
  loading: boolean;
}

function useFind<T = any, T2 = T>(
  type: string,
  args?: any,
  transformer?: (value: Array<T>) => Array<T2>
): [Array<T2>, Error | null, boolean] {
  const [{ value, error, loading }, setValue]: [
    IUseFind<T2>,
    (newValue: IUseFind<T2>) => void
  ] = React.useState({ value: [], error: null, loading: true } as any);
  const context = React.useContext(Context) as WorkerDBClientRX;

  function useFindChange(error: Error | null, value: any) {
    if (error) {
      setValue({ error, loading: false, value });
    }
    if (transformer) {
      setValue({ value: transformer(value), error: null, loading: false });
    } else {
      setValue({ value, error: null, loading: false });
    }
  }

  React.useEffect(
    context && context.query
      ? () => context.query(type, 'find', args, useFindChange)
      : () => {},
    [context, type, stringify(args)]
  );

  return [value, error, loading];
}

export default useFind;
