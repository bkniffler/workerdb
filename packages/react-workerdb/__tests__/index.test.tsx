import 'jest';
import * as wait from 'wait-for-expect';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { Find, WorkerDB, Call } from '../src';
import { inner } from '../../workerdb/src/worker';

describe('index', () => {
  it('should be able to be included', () => {
    require('../src/index');
  });

  it('should work', async () => {
    interface XType {
      _id: string;
      name: string;
    }

    const COLLECTION = { BIRD: 'bird' };
    const FindBird = Find.ofType<XType, Array<string>>();
    const CallBird = Call.ofType<XType, string>();

    let listener: any;
    listener = await getWorker(data => worker.onmessage({ data }));
    let ready = true;
    let error: any = null;
    const worker = {
      // postMessage: (data: any) => worker['onme' + 'ssage']({ data }),
      onmessage: (data: any) => data,
      postMessage: (data: any) => {
        listener(data);
      },
      terminate: () => { }
    };
    const component = renderer.create(
      <WorkerDB
        worker={worker}
        onReady={() => (ready = true)}
        onError={(err: Error) => (error = err)}
        adapter="memory"
      >
        <div>
          <FindBird
            collection={COLLECTION.BIRD}
            transform={data => data.map(({ name }) => name)}
            sort="name"
            render={data => (
              <React.Fragment>
                {data.map(x => (
                  <span key={x}>{x}</span>
                ))}
              </React.Fragment>
            )}
          />
          <CallBird
            collection={COLLECTION.BIRD}
            method="custom"
            render={data => <React.Fragment>{data}</React.Fragment>}
          />
        </div>
      </WorkerDB>
    );

    await (wait as any)(() => {
      expect(ready).toEqual(true);
      expect(error).toEqual(null);
    });
    expect(ready).toEqual(true);
    expect(error).toEqual(null);
    await new Promise(yay => setTimeout(yay, 3000));
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const getWorker = (cb: (data: any) => void): Function => {
  return inner(
    [
      {
        name: 'bird',
        schema: {
          type: 'object',
          version: 0,
          properties: {
            name: {
              type: 'string', index: true
            }
          }
        },
        methods: {
          custom: (col: any, value: any) =>
            col
              .findOne({ name: 'Filou' })
              .exec()
              .then((x: any) => x.name)
        }
      }
    ],
    cb,
    (rx: any) => {
      rx.plugin(require('pouchdb-adapter-memory'));
      return rx;
    },
    async (db: any) => {
      await db.bird.insert({ name: 'Oscar' });
      await db.bird.insert({ name: 'Oscar1' });
      await db.bird.insert({ name: 'Filou' });
      await db.bird.insert({ name: 'Oscar2' });
    }
  );
};
