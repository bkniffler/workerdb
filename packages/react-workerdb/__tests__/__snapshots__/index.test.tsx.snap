// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index should work 1`] = `
<div>
  <span>
    Filou
  </span>
  <span>
    Oscar
  </span>
  Filou
</div>
`;
